### PR TITLE
Persistence

### PR DESCRIPTION
**Pull request by evaj10**
Adds UnitOfWork mechanism for transaction handling.
Refactors repositories to use UnitOfWork.
Adds FK constraints and removes unused DbSets.
Refactors tests so that they don't persist changes to DB.

### CODE REVIEW

File: src/Tutor.Core/BuildingBlocks/Generics/ICrudRepository.cs [hunk 1]
```diff
@@ -9,5 +10,6 @@ public interface ICrudRepository<TEntity> where TEntity : Entity
     TEntity Create(TEntity entity);
     void BulkCreate(List<TEntity> entities);
     TEntity Update(TEntity entity);
-    void Delete(int id);
+    TEntity Update(TEntity storedEntity, TEntity entity);
+    Result Delete(int id);
```
Comment (Luburic) [inline src/Tutor.Core/BuildingBlocks/Generics/ICrudRepository.cs:14]: Should everything from the repository return Result or just void methods or just the Delete method?

File: src/Tutor.Core/Domain/Stakeholders/User.cs [hunk 2]
```diff
@@ -3,11 +3,11 @@
 public class User
 {
     public int Id { get; private set; }
-    public string Username { get; private set; }
+    public string Username { get; set; }
```
Comment (Luburic) [inline src/Tutor.Core/Domain/Stakeholders/User.cs:6]: Can this be `internal set`?

Btw, if you didn't already, install ReSharper.

File: src/Tutor.Core/UseCases/Learning/Assessment/SelectionService.cs [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ﻿using FluentResults;
+using System.Xml.Linq;
```
Comment (Luburic) [inline src/Tutor.Core/UseCases/Learning/Assessment/SelectionService.cs:2]: Is this an error?

File: src/Tutor.Core/UseCases/Learning/SessionService.cs [hunk 4]
```diff
@@ -26,6 +28,8 @@ public Result LaunchSession(int knowledgeComponentId, int learnerId)
 
         var result = kcMastery.LaunchSession();
         _knowledgeMasteryRepository.UpdateKcMastery(kcMastery);
+        var saveResult = _unitOfWork.Save();
```
Comment (Luburic) [inline src/Tutor.Core/UseCases/Learning/SessionService.cs:31]: You've the variable `result` in other classes, which is sufficient.

File: src/Tutor.Core/UseCases/Management/Enrollments/LearnerGroupService.cs [hunk 5]
```diff
@@ -32,12 +33,21 @@ public Result CreateMembers(int groupId, List<Learner> learners)
...
         return Result.Ok();
     }
 
     public Result DeleteMember(int groupId, int learnerId)
     {
-        _groupRepository.DeleteMember(groupId, learnerId);
+        var result = _groupRepository.DeleteMember(groupId, learnerId);
+        if (result.IsFailed) return result;
+
+        result = _unitOfWork.Save();
```
Comment (Luburic) [inline src/Tutor.Core/UseCases/Management/Enrollments/LearnerGroupService.cs:48]: To be clear, the general flow for using UoW is:

1. Invoke the correct repository method
    1. If the method in question is delete, check if the Result is failed.
2. Call UoW Save.
3. Check if its result is a failure.

1.1. stands out, as I've hinted at in one of the previous comments.

File: src/Tutor.Core/UseCases/Management/Stakeholders/CourseOwnershipService.cs [hunk 6]
```diff
@@ -29,7 +31,13 @@ public Result<Course> UpdateOwnedCourse(Course course, int instructorId)
         var isOwner = _ownedCourseRepository.IsCourseOwner(course.Id, instructorId);
         if (!isOwner) return Result.Fail(FailureCode.Forbidden);
 
-        return _courseRepository.Update(course);
+        _courseRepository.Update(course);
+        // DEMO: context caching (with and without save called)
```
Comment (Luburic) [inline src/Tutor.Core/UseCases/Management/Stakeholders/CourseOwnershipService.cs:35]: I imagine this is a reminder for our meeting, but pointing it out to avoid merging this code.

File: src/Tutor.Core/UseCases/Management/Stakeholders/StakeholderService.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        entity.UserId = user.Id;
+
+        _crudRepository.Update(dbStakeholder, entity);
+        user.Username = entity.Email;
+
+        var result = _unitOfWork.Save();
+        if (result.IsFailed) return result;
+
+        return entity;
+
```
Comment (Luburic) [inline src/Tutor.Core/UseCases/Management/Stakeholders/StakeholderService.cs:73]: Empty line.

File: src/Tutor.Infrastructure/Database/Repositories/CrudDatabaseRepository.cs [hunk 8]
```diff
@@ -32,29 +33,34 @@ public T Get(int id)
...
     {
-        DbContext.Entry(entity).State = EntityState.Modified;
-        DbContext.SaveChanges();
+        DbContext.Update(entity);
+        return entity;
+    }
+    public T Update(T storedEntity, T entity)
+    {
+        // DDD does not go well with _dbContext tracking changes if entity is present in context
+        // How it works: if entity is tracked, changes are registered when setter is called
```
Comment (Luburic) [inline src/Tutor.Infrastructure/Database/Repositories/CrudDatabaseRepository.cs:53]: Commenting to ensure we discuss this and then remove the comment.

File: src/Tutor.Infrastructure/Database/Repositories/LearningUtilities/FeedbackDatabaseRepository.cs [hunk 9]
```diff
@@ -14,12 +14,10 @@ public FeedbackDatabaseRepository(TutorContext dbContext)
     public void SaveEmotionsFeedback(EmotionsFeedback emotionsFeedback)
     {
         _dbContext.EmotionsFeedbacks.Attach(emotionsFeedback);
-        _dbContext.SaveChanges();
```
Comment (Luburic) [inline src/Tutor.Infrastructure/Database/Repositories/LearningUtilities/FeedbackDatabaseRepository.cs:17]: To avoid missing anything, did you do a search for "SaveChanges" and ensured that it is only present in UoW?

File: src/Tutor.Infrastructure/Database/TutorContext.cs [hunk 10]
```diff
@@ -72,24 +70,56 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
 
         modelBuilder.Entity<Stakeholder>().UseTpcMappingStrategy();
```
Comment (Luburic) [inline src/Tutor.Infrastructure/Database/TutorContext.cs:71]: 71-76 extract to ConfigureStakeholder

File: tests/Tutor.Web.Tests/Integration/Authoring/AssessmentTests.cs [hunk 11]
```diff
@@ -42,9 +42,11 @@ public void Creates()
             KnowledgeComponentId = -11,
             Text = "TT-1"
         };
+        dbContext.Database.BeginTransaction();
```
Comment (Luburic) [inline tests/Tutor.Web.Tests/Integration/Authoring/AssessmentTests.cs:45]: To discuss.

File: tests/Tutor.Web.Tests/Integration/Learning/Assessment/SelectionTests.cs [hunk 12]
```diff
@@ -70,6 +71,8 @@ public void Verifies_Mrq_has_no_feedback()
     {
         using var scope = Factory.Services.CreateScope();
         var controller = SetupAssessmentSelectionController(scope, "-2");
+        var dbContext = scope.ServiceProvider.GetRequiredService<TutorContext>();
+        dbContext.Database.BeginTransaction();
```
Comment (Luburic) [inline tests/Tutor.Web.Tests/Integration/Learning/Assessment/SelectionTests.cs:75]: Why is this needed for a read operation?

File: tests/Tutor.Web.Tests/Integration/Learning/Utilities/FeedbackTests.cs [hunk 13]
```diff
@@ -23,14 +23,15 @@ public void Stores_emotions_feedback(EmotionsFeedbackDto postedFeedback, Emotion
         using var scope = Factory.Services.CreateScope();
         var controller = SetupFeedbackController(scope);
         var dbContext = scope.ServiceProvider.GetRequiredService<TutorContext>();
+        dbContext.Database.BeginTransaction();
```
Comment (Luburic) [inline tests/Tutor.Web.Tests/Integration/Learning/Utilities/FeedbackTests.cs:26]: We should write down the guidelines this list https://app.clickup.com/24556453/v/l/li/900400270031 for using UoW in the context of production features and in the context of tests.

File: tests/Tutor.Web.Tests/Integration/Management/Courses/CourseCommandTests.cs [hunk 14]
```diff
@@ -27,16 +27,21 @@ public void Creates()
...
 
+        dbContext.ChangeTracker.Clear();
         result.ShouldNotBeNull();
         result.Id.ShouldNotBe(0);
         result.Code.ShouldBe(newEntity.Code);
         result.Name.ShouldBe(newEntity.Name);
         var storedEntity = dbContext.Courses.FirstOrDefault(i => i.Code == newEntity.Code);
         storedEntity.ShouldNotBeNull();
         storedEntity.Id.ShouldBe(result.Id);
+        var storedLearnerGroup = dbContext.LearnerGroups.FirstOrDefault(g => g.Name == "Group 1");
```
Comment (Luburic) [inline tests/Tutor.Web.Tests/Integration/Management/Courses/CourseCommandTests.cs:42]: Nice catch :)

**Review by Luburic — CHANGES_REQUESTED** [review]
Comments inline. Mostly requests for minor tweaks and a few comments to point out what we need to discuss.

We should also update the seed script to account for MRQ changes, but we can divide this work if there is too much to change.

File: src/Tutor.Core/Domain/Stakeholders/User.cs (hunk 2 above)
Comment (evaj10) [reply src/Tutor.Core/Domain/Stakeholders/User.cs:6]: It can, changed.

File: src/Tutor.Infrastructure/Database/Repositories/LearningUtilities/FeedbackDatabaseRepository.cs (hunk 9 above)
Comment (evaj10) [reply src/Tutor.Infrastructure/Database/Repositories/LearningUtilities/FeedbackDatabaseRepository.cs:17]: Yes, apart from EventStore repository, which is a different context that I didn't refactor.

**Review by Luburic — APPROVED** [review]
Everything looks great with the recent changes.
